blscosi/bdnproto/protocol_test.go

```diff
@@ -0,0 +1,97 @@
+package bdnproto
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.dedis.ch/cothority/v3"
+	"go.dedis.ch/cothority/v3/blscosi/protocol"
+	"go.dedis.ch/kyber/v3/pairing"
+	"go.dedis.ch/onet/v3"
+	"go.dedis.ch/onet/v3/log"
+)
+
+var testSuite = pairing.NewSuiteBn256()
+
+const testServiceName = "TestServiceBdnCosi"
+
+var testServiceID onet.ServiceID
+
+func init() {
+	id, err := onet.RegisterNewServiceWithSuite(testServiceName, testSuite, newService)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	testServiceID = id
+}
+
+func TestBdnProto_SimpleCase(t *testing.T) {
+	err := runProtocol(5, 1, 5)
+	require.NoError(t, err)
+
+	if !testing.Short() {
+		err = runProtocol(10, 5, 10)
+		require.NoError(t, err)
+
+		err = runProtocol(20, 5, 15)
+		require.NoError(t, err)
+	}
+}
+
+func runProtocol(nbrNodes, nbrSubTrees, threshold int) error {
+	local := onet.NewLocalTest(cothority.Suite)
+	defer local.CloseAll()
+	servers, roster, tree := local.GenTree(nbrNodes, false)
+
+	services := local.GetServices(servers, testServiceID)
+
+	rootService := services[0].(*testService)
+	pi, err := rootService.CreateProtocol(BdnProtocolName, tree)
+	if err != nil {
+		return err
+	}
+
+	cosiProtocol := pi.(*protocol.BlsCosi)
+	cosiProtocol.CreateProtocol = rootService.CreateProtocol
+	cosiProtocol.Msg = []byte{0xFF}
+	cosiProtocol.Timeout = 10 * time.Second
+	cosiProtocol.Threshold = threshold
+	if nbrSubTrees > 0 {
+		err = cosiProtocol.SetNbrSubTree(nbrSubTrees)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = cosiProtocol.Start()
+	if err != nil {
+		return err
+	}
+
+	select {
+	case sig := <-cosiProtocol.FinalSignature:
+		pubs := roster.ServicePublics(testServiceName)
```

Comment:
Please verify that `BdnSignature(sig).Verify(testSuite, theWrongMessage, pubs)` returns an error. You might need to pass the *testing.T into runProtocol to call `require.Error(t, ...)`.